File: lib/motion.dart
import 'package:vector_math/vector_math_64.dart';

/// A Motion can be used to define a Moveable entity with a Position.
class Motion {
  Vector2 start;
  int startUsec;
  // Speed of the entity in m/sec
  double speed;
  Vector2 dest;

  Motion(this.start, this.startUsec, this.speed, this.dest);

  factory Motion.idle(Vector2 pos, int currUsec) {
    return Motion(pos, currUsec, 0, pos);
  }

  Vector2 getStart() {
    return start;
  }

  Vector2 currentPosition(int currUsec) {
    if (start.distanceTo(dest) < 0.001) {
      return dest;
    }

    Vector2 distVec = dest - start;
    Vector2 dirVec = distVec.normalized();
    Vector2 speedVec = dirVec * speed;

    double deltaSec = (currUsec - startUsec) / 1000000.0;
    Vector2 currPos = start + speedVec * deltaSec;

    if (distVec.length2 < currPos.length2) {
      return dest;
    }
    return currPos;
  }

  Motion moveTo(int currUsec, double speed, Vector2 dest) {
    Vector2 currPos = currentPosition(currUsec);
    return Motion(currPos, currUsec, speed, dest);
  }
}

File: server/cozyworld/entity/motion.go
package entity

import (
	"github.com/go-gl/mathgl/mgl64"
	cpb "github.com/le-michael/cozyworld/protos"
)

// A Motion can be used to define a Moveable entity with a Position.
type Motion struct {
	start     mgl64.Vec2
	startUsec int64
	// Speed of the entity in m/sec
	speed float64
	dest  mgl64.Vec2
}

func NewIdleMotion(pos mgl64.Vec2, currUsec int64) Motion {
	return NewMotion(pos, currUsec, 0, pos)
}

func NewMotion(start mgl64.Vec2, startUsec int64, speed float64, dest mgl64.Vec2) Motion {
	return Motion{
		start:     start,
		startUsec: startUsec,
		speed:     speed,
		dest:      dest,
	}
}

func (m Motion) Start() mgl64.Vec2 {
	return m.start
}

func (m Motion) CurrentPosition(currUsec int64) mgl64.Vec2 {
	if m.start.ApproxEqual(m.dest) {
		return m.dest
	}

	distVec := m.dest.Sub(m.start)
	dirVec := distVec.Normalize()
	speedVec := dirVec.Mul(m.speed)

	deltaSec := float64((currUsec - m.startUsec)) / 1000000.0
	currPos := m.start.Add(speedVec.Mul(deltaSec))

	if distVec.LenSqr() < currPos.LenSqr() {
		return m.dest
	}
	return currPos
}

func (m Motion) MoveTo(currUsec int64, speed float64, dest mgl64.Vec2) Motion {
	currPos := m.CurrentPosition(currUsec)
	return Motion{
		start:     currPos,
		startUsec: currUsec,
		speed:     speed,
		dest:      dest,
	}
}

func (m Motion) ToProto() *cpb.Motion {
	return &cpb.Motion{
		Start:     &cpb.Vec2{X: m.start.X(), Y: m.start.Y()},
		StartUsec: m.startUsec,
		Speed:     m.speed,
		Dest:      &cpb.Vec2{X: m.dest.X(), Y: m.dest.Y()},
	}
}

func (m *Motion) FromProto(pb *cpb.Motion) {
	m.start = mgl64.Vec2{pb.Start.X, pb.Start.Y}
	m.startUsec = pb.StartUsec
	m.speed = pb.Speed
	m.dest = mgl64.Vec2{pb.Dest.X, pb.Dest.Y}
}
